### PR TITLE
add prepend!

### PR DESCRIPTION
**Pull request by bkamins**
Another step towards https://github.com/JuliaData/DataFrames.jl/issues/2936

### CODE REVIEW

File: src/dataframe/dataframe.jl [hunk 1]
```diff
@@ -1387,24 +1476,42 @@ function Base.append!(df1::DataFrame, df2::AbstractDataFrame; cols::Symbol=:sete
...
                 end
             else
                 if Missing <: eltype(df1[!, j])
-                    resize!(df1[!, j], targetrows)
-                    df1[nrows+1:targetrows, j] .= missing
+                    if atend
+                        resize!(df1[!, j], targetrows)
+                        df1[nrow1+1:targetrows, j] .= missing
+                    else
+                        prepend!(df1[!, j], Iterators.repeated(missing, nrow2))
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:1502]: Why not use the same approach base don `resize!` as elsewhere?
```

File: src/dataframe/dataframe.jl [hunk 2]
```diff
@@ -1422,15 +1529,28 @@ function Base.append!(df1::DataFrame, df2::AbstractDataFrame; cols::Symbol=:sete
...
         # Undo changes in case of error
         for col in _columns(df1)
-            resize!(col, nrows)
+            @assert length(col) >= nrow1
+            if atend
+                resize!(col, nrow1)
+            else
+                if length(col) != nrow1
+                    deleteat!(col, 1:length(col) - nrow1)
+                end
```
Comment (nalimilan) [inline src/dataframe/dataframe.jl:1552]:
```suggestion
            elseif length(col) != nrow1
                deleteat!(col, 1:length(col) - nrow1)
```

File: test/dataframe.jl [hunk 3]
```diff
@@ -596,6 +596,44 @@ end
...
+    for df in [DataFrame(a=Any[1]), DataFrame(a=1)]
+        @test push!(df, (b=1,), cols=:union) ≅
+              DataFrame(a=[1, missing], b=[missing, 1])
+        @test push!(df, (b=1,), cols=:union) ≅
+              DataFrame(a=[1, missing, missing], b=[missing, 1, 1])
+        df.x = 1:3
+        with_logger(SimpleLogger(IOBuffer())) do
+            @test_throws MethodError push!(df, (b=1,), cols=:union, promote=false)
+        end
+        @test df  ≅ DataFrame(a=[1, missing, missing], b=[missing, 1, 1], x=1:3)
```
Comment (nalimilan) [inline test/dataframe.jl:609]:
```suggestion
        @test df ≅ DataFrame(a=[1, missing, missing], b=[missing, 1, 1], x=1:3)
```

File: test/dataframe.jl [hunk 4]
```diff
@@ -596,6 +596,44 @@ end
...
+        @test push!(df, DataFrame(b=1)[1, :], cols=:union) ≅
+              DataFrame(a=[1, missing], b=[missing, 1])
+        @test push!(df, DataFrame(b=1)[1, :], cols=:union) ≅
+              DataFrame(a=[1, missing, missing], b=[missing, 1, 1])
+        df.x = 1:3
+        with_logger(SimpleLogger(IOBuffer())) do
+            @test_throws MethodError push!(df, DataFrame(b=1)[1, :], cols=:union,
+                                           promote=false)
+        end
+        @test df  ≅ DataFrame(a=[1, missing, missing], b=[missing, 1, 1], x=1:3)
```
Comment (nalimilan) [inline test/dataframe.jl:626]:
```suggestion
        @test df ≅ DataFrame(a=[1, missing, missing], b=[missing, 1, 1], x=1:3)
```

**Review by nalimilan — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**